4-WebApp-your-API/4-3-AnyOrg/TodoListService/Startup.cs

```diff
@@ -25,43 +26,43 @@ public Startup(IConfiguration configuration)
 
-            services.AddMicrosoftIdentityWebApiAuthentication(Configuration)
-                    .EnableTokenAcquisitionToCallDownstreamApi()
-                    .AddInMemoryTokenCaches();
-
-            // Comment above lines of code and uncomment this section if you would like to validate ID tokens for allowed tenantIds
-            //services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
-            //  .AddMicrosoftIdentityWebApi(options =>
-            //{
-            //    Configuration.Bind("AzureAd", options);
-            //    options.Events = new JwtBearerEvents();
-            //    options.Events.OnTokenValidated = async context =>
-            //    {
-            //        string[] allowedTenants = {/* list of tenant IDs */ };
-            //        string tenantId = context.Principal.Claims.FirstOrDefault(x => x.Type == "tid" || x.Type == "http://schemas.microsoft.com/identity/claims/tenantid")?.Value;
+            //get list of allowed tenants from configuration
+            var allowed = Configuration.GetSection("AzureAd:AllowedTenants").Get<string[]>();
+            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
+              .AddMicrosoftIdentityWebApi(options =>
+            {
+                Configuration.Bind("AzureAd", options);
+                options.Events = new JwtBearerEvents();
+                options.Events.OnTokenValidated = async context =>
+                {
+                    await Task.Run(() =>
+                    {
+                        string[] allowedTenants = allowed;
+                        string tenantId = context.Principal.Claims.FirstOrDefault(x => x.Type == ClaimConstants.Tid || x.Type == ClaimConstants.TenantId)?.Value;
 
-            //        if (!allowedTenants.Contains(tenantId))
-            //        {
-            //            throw new Exception("This tenant is not authorized");
-            //        }
-            //    };
-            //}, options => { Configuration.Bind("AzureAd", options); })
-            //  .EnableTokenAcquisitionToCallDownstreamApi(
-            //        options =>
-            //        {
-            //             Configuration.Bind("AzureAd", options);
-            //        })
-            //    .AddInMemoryTokenCaches();
+                        if (!allowedTenants.Contains(tenantId))
+                        {
+                            throw new UnauthorizedAccessException("This tenant is not authorized");
+                        }
+                    });
+                };
+            }, options => { Configuration.Bind("AzureAd", options); })
+              .EnableTokenAcquisitionToCallDownstreamApi(
+                    options =>
+                    {
+                        Configuration.Bind("AzureAd", options);
+                    })
+              .AddInMemoryTokenCaches();
 
             // Creating policies that wraps the authorization requirements
             services.AddAuthorization();
```

Comment:
Another possibility would be to use a policy
That would be my preference. See

https://github.com/AzureAD/microsoft-identity-web/wiki/authorization-policies#filtering-tenants